odoo/tools/profiler.py

```diff
@@ -211,49 +204,58 @@ def start(self):
-
+        while self._end_time is None:
+            self.add()
+            self.sleep()
         self._entries.append({'stack': [], 'start': real_time()})  # add final end frame
 
+    def sleep(self):
+        # Note: This may not be very precise. Most systems will sleep at
+        # minimum between 1 and 5 ms. "The suspension time may be longer
+        # than requested by an arbitrary amount, because of the scheduling
+        # of other activity in the system."
+        time.sleep(self.frame_interval)
+
     def stop(self):
-        self.active = False
+        self._end_time = real_time()
         self.__thread.join()
-        self.profiler.init_thread.profile_hooks.remove(self.progress)
+        self.profiler.init_thread.profile_hooks.remove(self.add)
 
 
 class PeriodicCollector(_BasePeriodicCollector):
 
     name = 'traces_async'
 
-    def add(self, entry=None, frame=None):
+    def __init__(self, interval=None):
+        super().__init__(interval)
+        self._last_frame_id = 0  # keep the id of the object to avoid keeping references
+        self._last_time = 0
+
+    def add(self, entry=None, frame=None, check_limit=True):
         """ Add an entry (dict) to this collector. """
-        if self.last_frame:
-            duration = real_time() - self._last_time
-            if duration > self.frame_interval * 10 and self.last_frame:
-                # The profiler has unexpectedly slept for more than 10 frame intervals. This may
-                # happen when calling a C library without releasing the GIL. In that case, the
-                # last frame was taken before the call, and the next frame is after the call, and
-                # the call itself does not appear in any of those frames: the duration of the call
-                # is incorrectly attributed to the last frame.
-                self._entries[-1]['stack'].append(('profiling', 0, '⚠ Profiler freezed for %s s' % duration, ''))
-            self.last_frame = None  # skip duplicate detection for the next frame.
-        self._last_time = real_time()
+        now = real_time()
+        if self._last_frame_id and (duration := now - self._last_time) > self.frame_interval * 10:
+            # The profiler has unexpectedly slept for more than 10 frame intervals. This may
+            # happen when calling a C library without releasing the GIL. In that case, the
+            # last frame was taken before the call, and the next frame is after the call, and
+            # the call itself does not appear in any of those frames: the duration of the call
+            # is incorrectly attributed to the last frame.
+            self._entries[-1]['stack'].append(('profiling', 0, '⚠ Profiler freezed for %s s' % duration, ''))
+        self._last_time = now
 
         frame = frame or get_current_frame(self.profiler.init_thread)
-        if frame == self.last_frame:
+        frame_id = id(frame)
+        if frame_id == self._last_frame_id:
```

Comment:
Issue: if the former frame has been deleted, `self._last_frame_id` may now be the address of another object, and potentially another frame...  The `id()` of an object is a valid identifier _while the object is alive_.

You may use a `weakref` instead.  When the object is deleted, the weakref refers to `None` and the obviously you no longer are in the frame, since the frame is gone.